File: Projects/Project_06_2D_Wave/For_Students/Sample_Code_Base/compute_normals.m
%----------------------------------------------------------------------%
%This subroutine builds the Normals for a 
%CGDG Method on  Quads
%Written by Francis X. Giraldo
%           Naval Postgraduate School
%           Department of Applied Mathematics
%           Monterey, CA 93943-5216
%----------------------------------------------------------------------%
function [normals,jac_face]=compute_normals(face,intma,coord,...
                          nside,ngl,nq,psi,dpsi)
%global arrays
normals=zeros(2,nq,nside); 
jac_face=zeros(nq,nside);

%local arrays
x=zeros(ngl,ngl);
y=zeros(ngl,ngl);

%loop thru the sides
for is=1:nside

    %Store Left and Right Elements
    ilocl=face(1,is);
    ilocr=face(2,is);
    iel=face(3,is);
    ier=face(4,is);

    %Store Element Variables
    for j=1:ngl
    for i=1:ngl
        ip=intma(i,j,iel);
        x(i,j)=coord(1,ip);
        y(i,j)=coord(2,ip);
    end %i  
    end %j  

    %Construct Mapping Derivatives: dx/dksi, dx/deta, dy/dksi,
    %dy/deta
    [x_ksi,x_eta]=map_deriv(psi,dpsi,x,ngl,nq);
    [y_ksi,y_eta]=map_deriv(psi,dpsi,y,ngl,nq);
    
    %Compute Normals 
    for l=1:nq
         switch (ilocl)
            case (1)
               %Side 1: eta=-1
               i=l;
               j=1;
               normals(1,l,is)=+y_ksi(i,j);
               normals(2,l,is)=-x_ksi(i,j);
               jac_face(l,is)=sqrt( x_ksi(i,j)^2 + y_ksi(i,j)^2 );
            case (2)
               %Side 2: ksi=+1
               i=nq;
               j=l;
               normals(1,l,is)=+y_eta(i,j);
               normals(2,l,is)=-x_eta(i,j);
               jac_face(l,is)=sqrt( x_eta(i,j)^2 + y_eta(i,j)^2 );
            case (3)
               %Side 3: eta=+1
               i=nq+1-l;
               j=nq;
               normals(1,l,is)=-y_ksi(i,j);
               normals(2,l,is)=+x_ksi(i,j);
               jac_face(l,is)=sqrt( x_ksi(i,j)^2 + y_ksi(i,j)^2 ); 
            case (4)
               %Side 4: ksi=-1
               i=1;
               j=nq+1-l;
               normals(1,l,is)=-y_eta(i,j);
               normals(2,l,is)=+x_eta(i,j);
               jac_face(l,is)=sqrt( x_eta(i,j)^2 + y_eta(i,j)^2 );
         end %switch 
    end %l   

    %Normalize Normals
    for l=1:nq
        rnx=sqrt( normals(1,l,is)^2 + normals(2,l,is)^2 );
        normals(1,l,is)=normals(1,l,is)/rnx;
        normals(2,l,is)=normals(2,l,is)/rnx;
     end %l   

end %is

File: Projects/Project_06_2D_Wave/For_Students/julia/compute_normals.jl
#=
---------------------------------------------------------------------
This function computes the normal vectors and Jacobian of each FACE.

Written by F.X. Giraldo on July 14, 2021
           Department of Applied Mathematics
           Naval Postgraduate School
           Monterey; CA 93943-5216
---------------------------------------------------------------------
=#

include("map_deriv.jl")

function compute_normals(face,intma,coord,Nface,Np,Nq,ψ,dψ)

    #global arrays
    normals=zeros(DFloat,2,Nq,Nface)
    jac_face=zeros(DFloat,Nq,Nface)

    #local arrays
    x=zeros(DFloat,Np,Np)
    y=zeros(DFloat,Np,Np)

    #loop thru the sides
    for f=1:Nface

        #Store Left & Right Elements
        ilocl=face[1,f]
        ilocr=face[2,f]
        iel=face[3,f]
        ier=face[4,f]

        #Store Element Variables
        for j=1:Np, i=1:Np
            I=intma[i,j,iel]
            x[i,j]=coord[1,I]
            y[i,j]=coord[2,I]
        end #i, j
 
        #Construct Mapping Derivatives: dx/dksi; dx/deta; dy/dksi
        #dy/deta
        (x_ξ,x_η)=map_deriv(ψ,dψ,x,Np,Nq,DFloat)
        (y_ξ,y_η)=map_deriv(ψ,dψ,y,Np,Nq,DFloat)

        #Compute Normals 
        for l=1:Nq
            if (ilocl == 1) 
                #Side 1: eta=-1
                i=l
                j=1
                normals[1,l,f]=+y_ξ[i,j]
                normals[2,l,f]=-x_ξ[i,j]
                jac_face[l,f]=sqrt( x_ξ[i,j]^2 + y_ξ[i,j]^2 )
            elseif (ilocl == 2) 
                #Side 2: ksi=+1
                i=Nq
                j=l
                normals[1,l,f]=+y_η[i,j]
                normals[2,l,f]=-x_η[i,j]
                jac_face[l,f]=sqrt( x_η[i,j]^2 + y_η[i,j]^2 )
            elseif (ilocl == 3)
                #Side 3: eta=+1
                i=Nq+1-l
                j=Nq
                normals[1,l,f]=-y_ξ[i,j]
                normals[2,l,f]=+x_ξ[i,j]
                jac_face[l,f]=sqrt( x_ξ[i,j]^2 + y_ξ[i,j]^2 )
            elseif (ilocl == 4)
                #Side 4: ksi=-1
                i=1
                j=Nq+1-l
                normals[1,l,f]=-y_η[i,j]
                normals[2,l,f]=+x_η[i,j]
                jac_face[l,f]=sqrt( x_η[i,j]^2 + y_η[i,j]^2 )
            end
        end #l 

        #Normalize Normals
        for l=1:Nq
            rnx=sqrt( normals[1,l,f]^2 + normals[2,l,f]^2 )
            normals[1,l,f]=normals[1,l,f]/rnx
            normals[2,l,f]=normals[2,l,f]/rnx
        end #l   

    end #s
    
    return(normals,jac_face)
end
